sdk/cosmos/azure_data_cosmos/src/location_cache.rs

```diff
@@ -0,0 +1,570 @@
+                ),
+                (
+                    "Location 2".to_string(),
+                    "https://location2.documents.azure.com".to_string()
+                ),
+                (
+                    "Location 3".to_string(),
+                    "https://location3.documents.azure.com".to_string()
+                ),
+                (
+                    "Location 4".to_string(),
+                    "https://location4.documents.azure.com".to_string()
+                )
+            ])
+        );
+
+        assert_eq!(
+            cache.locations_info.write_endpoints,
+            vec![
+                "https://location1.documents.azure.com".to_string(),
+                "https://location2.documents.azure.com".to_string()
+            ]
+        );
+
+        assert_eq!(
+            cache.locations_info.read_endpoints,
+            vec![
+                "https://location1.documents.azure.com".to_string(),
+                "https://location2.documents.azure.com".to_string(),
+            ]
+        );
+    }
+
+    #[test]
+    fn location_cache_update_with_one_preferred_location() {
+        let (default_endpoint, write_locations, read_locations, preferred_locations) =
+            create_test_data();
+
+        let mut cache = LocationCache::new(
+            default_endpoint.to_string(),
+            vec![preferred_locations[0].clone()],
+        );
+
+        cache.update(write_locations, read_locations);
+
+        assert_eq!(
+            cache.default_endpoint,
+            "https://default.documents.azure.com"
+        );
+
+        assert_eq!(
+            cache.locations_info.preferred_locations,
+            vec![preferred_locations[0].clone()]
+        );
+
+        assert_eq!(
+            cache.locations_info.write_endpoints,
+            vec!["https://location1.documents.azure.com".to_string()]
+        );
+
+        assert_eq!(
+            cache.locations_info.read_endpoints,
+            vec!["https://location1.documents.azure.com".to_string()]
+        );
+    }
+
+    #[test]
+    fn mark_read_endpoint_unavailable_test() {
+        // set up test data
+        let (default_endpoint, write_locations, read_locations, preferred_locations) =
+            create_test_data();
+
+        let mut cache = LocationCache::new(default_endpoint, preferred_locations);
+        cache.update(write_locations, read_locations);
+
+        // mark location 1 as unavailable endpoint for read operation
+        let unavailable_endpoint = "https://location1.documents.azure.com";
+        let operation = RequestOperation::Read;
+        cache.mark_endpoint_unavailable(unavailable_endpoint, operation);
+
+        // check that endpoint is last option in read endpoints and it is in the location unavailability info map
+        assert_eq!(
+            cache.locations_info.read_endpoints.last(),
+            Some(&unavailable_endpoint.to_string())
+        );
```

Comment:
I'd suggest asserting the entire content of the list again, since you control the test data. You know that it should be `location2` then `location1`, so rather than just checking for `location1` at the end, check the whole list. That way, your test also catches bugs like `location2` getting dropped as well (through some unintentional issue elsewhere)